lib/runtime/wazero/imports.go

```diff
@@ -0,0 +1,2247 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package wazero_runtime
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"reflect"
+	"time"
+
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/ChainSafe/gossamer/lib/crypto/secp256k1"
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+	"github.com/ChainSafe/gossamer/lib/runtime"
+	"github.com/ChainSafe/gossamer/lib/transaction"
+	"github.com/ChainSafe/gossamer/lib/trie"
+	"github.com/ChainSafe/gossamer/lib/trie/proof"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/tetratelabs/wazero/api"
+)
+
+var (
+	logger = log.NewFromGlobal(
+		log.AddContext("pkg", "runtime"),
+		log.AddContext("module", "wazero"),
+	)
+)
+
+const (
+	validateSignatureFail = "failed to validate signature"
+)
+
+// toPointerSize converts an uint32 pointer and uint32 size
+// to an int64 pointer size.
+func newPointerSize(ptr, size uint32) (pointerSize uint64) {
+	return uint64(ptr) | (uint64(size) << 32)
+}
+
+// splitPointerSize converts a 64bit pointer size to an
+// uint32 pointer and a uint32 size.
+func splitPointerSize(pointerSize uint64) (ptr, size uint32) {
+	return uint32(pointerSize), uint32(pointerSize >> 32)
+}
+
+// read will read from 64 bit pointer size and return a byte slice
+func read(m api.Module, pointerSize uint64) (data []byte) {
+	ptr, size := splitPointerSize(pointerSize)
+	data, ok := m.Memory().Read(ptr, size)
+	if !ok {
+		panic("write overflow")
+	}
+	return data
+}
+
+// copies a Go byte slice to wasm memory and returns the corresponding
+// 64 bit pointer size.
+func write(m api.Module, allocator *runtime.FreeingBumpHeapAllocator, data []byte) (pointerSize uint64, err error) {
+	size := uint32(len(data))
+	pointer, err := allocator.Allocate(size)
+	if err != nil {
+		return 0, fmt.Errorf("allocating: %w", err)
+	}
+
+	ok := m.Memory().Write(pointer, data)
+	if !ok {
+		return 0, fmt.Errorf("out of range")
+	}
```

Comment:
Do we not need to copy memory here any more?

We used to do:
`copy(memory[ptr:ptr+size], data)
pointerSize = toPointerSize(ptr, size)`
	
Why do we not need this copy any more?

Reply:
We are using `api.Memory` interface provided by `wazero`.  This isn't needed anymore since it's guarding against accessing the underlying memory which was pretty unsafe to begin with.